### PR TITLE
Customized Parsers

### PR DESCRIPTION
**Pull request by mrh1997**
Removed all global variables from Parser implementation to allow working with multiple independent Parser implementations in the same project. For this purpose some major API refactorings were done:

* one Parser **class** per C dialect (currently only ANSI C and MSVC are supported)
* one Parser **object** per environment (=parameters usually provided to compiler via command line, i.e. include directories, preprocessor defines)
* the Parser can process multiple header files via .read(). Every invokation of read adds the corresponding declarations to the CLibInterface which is used as backend.
* .read() is composed of multiple other methods (.remove_comments(), .preprocess(), .parse()), which can be invoked separately.
* every .read() throws away the source code after processing it and storing the results to clib_intf (=> saves memory)

### CODE REVIEW

File: pyclibrary/c_parser.py [hunk 1]
```diff
@@ -34,22 +34,29 @@
      WordEnd, Keyword, Regex, Literal, SkipTo, ZeroOrMore, OneOrMore,
      Group, LineEnd, stringStart, quotedString, oneOf, nestedExpr,
      delimitedList, restOfLine, cStyleComment, alphas, alphanums, hexnums,
-     lineno, Suppress)
+     lineno, Suppress, NoMatch, ParseException)
 ParserElement.enablePackrat()
 
 logger = logging.getLogger(__name__)
 
+__all__ = ['win_defs', 'CParser', 'MSVCParser', 'SYS_HEADER_DIRS']
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: This should be win_defs_parser rather than win_defs, no ?

File: pyclibrary/c_parser.py [hunk 2]
```diff
@@ -60,193 +67,212 @@ def win_defs(version='1500'):
...
-        process_all=False
-        )
+    if sdk_dir is None:
+        sdk_dir = []
+    dir = os.path.dirname(__file__)
+    cache_file_name = os.path.join(dir, 'headers', 'WinDefs.cache')
+
+    parser = MSVCParser(header_dirs=[sdk_dir],
+                        predef_macros={'_WIN32': '', 'NO_STRICT': ''},
+                        msc_ver=version, arch=32, )
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: Does the architecture make any differences for the windows headers ?

File: pyclibrary/c_parser.py [hunk 3]
```diff
@@ -60,193 +67,212 @@ def win_defs(version='1500'):
...
+
+    # the fix header file order is very fragile, as there is no clean
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: Nice catch !

File: pyclibrary/c_parser.py [hunk 4]
```diff
@@ -60,193 +67,212 @@ def win_defs(version='1500'):
...
+            self.cur_file_name = None
+
+    def reset_clib_intf(self):
+        """Detaches the current .clib_intf object from the parser and creates
+        a new one.
 
-        if cache is not None:
-            logger.debug("Writing cache file '{}'".format(cache))
-            self.write_cache(cache)
+        Returns
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: This function does not return anything currently.

File: pyclibrary/c_parser.py [hunk 5]
```diff
@@ -267,11 +293,10 @@ def load_cache(self, cache_file, check_validity=False):
 
         Returns
         -------
-        result : bool
-            Did the loading succeeded.
+        result : CLibInterface
+            cache lib interface read from interface.
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: 'Lib interface read from the cache file' rather, no ?

File: pyclibrary/c_parser.py [hunk 6]
```diff
@@ -527,26 +523,27 @@ def pa(t):
...
-                        self.clib_intf.del_macro(macroName.strip())
-                    except Exception:
-                        if sys.exc_info()[0] is not KeyError:
-                            mess = "Error removing macro definition '{}'"
-                            logger.exception(mess.format(macroName.strip()))
+                    if if_true[-1]:
+                        try:
+                            self.clib_intf.del_macro(macro_name.strip())
+                        except Exception:
+                            if sys.exc_info()[0] is not KeyError:
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: I would prefer 
```python
except KeyError:
    pass
except Exception:
    # custom message here ```

File: pyclibrary/c_parser.py [hunk 7]
```diff
@@ -658,7 +654,9 @@ def expand_macros(self, line):
         Faulty calls to macro function are left untouched.
 
         """
-        reg = re.compile(r'("(\\"|[^"])*")|(\b(\w+)\b)')
+        reg = re.compile(
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: Could you add a comment on this regex ?

File: pyclibrary/c_parser.py [hunk 8]
```diff
@@ -735,20 +737,202 @@ def parse_defs(self, path, return_unparsed=False):
...
+                         list(self.filter_par(keywords)))
+
+    def generic_ident(self, exceptions=None):
+        """A generic ParserElement for identificators. In C this are all words
+        not beginning with numbers.
+        The ParserElement 'exception' allows to specify a list of not
+        accepted identificators.
+
+        Can be overwritten by descendant to allow only specific ID patterns.
+        """
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: missing blank line at the end of docstring.

File: pyclibrary/c_parser.py [hunk 9]
```diff
@@ -1139,226 +1336,61 @@ def eval_expr(self, toks):
...
+    predef_macros : dict[str, str], optional
+        same as CParser
 
-    expression << Group(atom + ZeroOrMore(bi_operator + atom))
-    expression.setParseAction(converter(str))
+    msc_ver : int, optional
+        The C compiler version that shall be emulated (usually 1500 for
+        version 15.00)
 
+    arch : int, optional
```
Comment (MatthieuDartiailh) [inline pyclibrary/c_parser.py]: you can specify the values writing ```arch : {32, 64}, optional```

File: tests/headers/file_handling/cpp_comments.h [hunk 10]
```diff
@@ -1,6 +1,6 @@
 // This is file is used to test the removal of C++ style comments
 
-# define MACRO 1 // inline comment
+# define MACRO 2 // inline comment
```
Comment (MatthieuDartiailh) [inline tests/headers/file_handling/cpp_comments.h]: Why the change ?

File: tests/headers/structs.h [hunk 11]
```diff
@@ -29,7 +29,7 @@ struct recursive_struct {
 typedef struct tagWNDCLASSEXA {
     int         cbClsExtra;
     int         cbWndExtra;
-} WNDCLASSEXA, *PWNDCLASSEXA, near *NPWNDCLASSEXA, far *LPWNDCLASSEXA;
```
Comment (MatthieuDartiailh) [inline tests/headers/structs.h]: Are far and near tested somewhere else ? or do you have a reason to remove support for them ?

File: tests/test_parser.py [hunk 12]
```diff
@@ -13,14 +13,16 @@
...
 
 import os
-import sys
+import tempfile
+import shutil
+import io
 from pytest import raises
-from pyclibrary.c_parser import CParser
-import pyclibrary.utils
+from pyclibrary.c_parser import CParser, InvalidCacheError, MSVCParser
```
Comment (MatthieuDartiailh) [inline tests/test_parser.py]: Why importing the InvalidCache error from c_parser rather than errors ?

File: tests/test_parser.py [hunk 13]
```diff
@@ -36,104 +38,167 @@ class TestFileHandling(object):
...
+        assert clib_intf['MACRO_B'].content == 'base_val_b'
+
+        parser.reset_clib_intf()
+        assert parser.file_order == []
+        parser.read(self.hdr_file_path('multi_file3.h'))
+        assert parser.clib_intf['MACRO_A'].content == 'other_val_a'
+        assert 'MACRO_B' not in parser.clib_intf
+        assert clib_intf['MACRO_A'].content == 'replaced_val_a'
+
+    def test_caching(self):
```
Comment (MatthieuDartiailh) [inline tests/test_parser.py]: here use the tmpdir fixture from pytest (simply add the argument to the function and use tmpdir as any directory path). It avoids messing with tempfile and shutil.

**Comment by MatthieuDartiailh** [comment]
First review done. Great work as always !

File: pyclibrary/c_parser.py (hunk 1 above)
Comment (mrh1997) [inline pyclibrary/c_parser.py]: Oh. Missed this line, when changing the name to 'win_defs_parser'. Next commit will fix this.

File: pyclibrary/c_parser.py (hunk 2 above)
Comment (mrh1997) [inline pyclibrary/c_parser.py]: Unfortunately: Yes!

There are some ```#ifdefs``` in the windows headers, that check for one of the architecture defines (```_M_IX86```, ```_M_AMD64```, ...). Honsestly said, I do not know if the difference is relevant for us, but I could imagine, that there are corner cases, where it **is** relevant if you specified a platform define or not.

In the next commit the architecture will even be configurable (32 or 64 bit), although in most cases this should make no difference...

File: pyclibrary/c_parser.py (hunk 6 above)
Comment (mrh1997) [inline pyclibrary/c_parser.py]: Of course!!!
Don't know how I could oversee that code smell when modifying the code.

File: pyclibrary/c_parser.py (hunk 4 above)
Comment (mrh1997) [inline pyclibrary/c_parser.py]: done...

File: pyclibrary/c_parser.py (hunk 7 above)
Comment (mrh1997) [inline pyclibrary/c_parser.py]: done...

File: pyclibrary/c_parser.py (hunk 8 above)
Comment (mrh1997) [inline pyclibrary/c_parser.py]: done...

File: pyclibrary/c_parser.py (hunk 9 above)
Comment (mrh1997) [inline pyclibrary/c_parser.py]: done...

File: tests/headers/structs.h (hunk 11 above)
Comment (mrh1997) [inline tests/headers/structs.h]: Neither __far, __near nor far, near are supported by up-to-date visual compiler versions.
This is a legacy from the days when compilers generated 16bit code.

If you have a look into the windows header files, you'll see, that they define these identifiers to nothing:
```C
#define far
```

File: tests/test_parser.py (hunk 12 above)
Comment (mrh1997) [inline tests/test_parser.py]: fixed it...

File: tests/test_parser.py (hunk 13 above)
Comment (mrh1997) [inline tests/test_parser.py]: nice hint.
Didn't know about this pytest feature, yet...

File: tests/headers/file_handling/cpp_comments.h (hunk 10 above)
Comment (mrh1997) [inline tests/headers/file_handling/cpp_comments.h]: During development of f3ed1ba1c2671391df94d31940f048828bd725cb I used cpp_comments.h for test_process_multiple_files. Thus I needed different values for the ```MACRO``` macro to check if the last parsed file is overwriting the macro definitions of the earlier files.

But then I reworked this by providing extra files for this test, instead of misusing ```cpp_comments.h``` without undoing this (then unnecessary change)...

**Comment by MatthieuDartiailh** [comment]
Merging as all comments have been addressed.
